std/machines/hash/poseidon2_bb.asm

```diff
@@ -0,0 +1,288 @@
+    // part of the state, so we can skip the corresponding rows in the matrix.
+    let apply_mds = |input, output_len| {
+        let dot_product = |v1, v2| array::sum(array::zip(v1, v2, |v1_i, v2_i| v1_i * v2_i));
+        array::map(array::sub_array(MDS, 0, output_len), |row| dot_product(row, input))
+    };
+
+    let external_round = constr |c_idx, input, output| {
+        // Add constants
+        let step_a = array::zip(input, EXTERNAL_ROUND_CONSTANTS[c_idx], |v, c| v + c);
+
+        // Apply S-box
+        let x3 = array::map(step_a, constr |x| { let x3; x3 = x * x * x; x3});
+        let x7 = array::zip(x3, step_a, constr |x_3, x| { let x7; x7 = x_3 * x_3 * x; x7 });
+
+        // Multiply with MDS Matrix
+        array::zip(output, apply_mds(x7, array::len(output)), |out, x| out = x);
+    };
+
+    let internal_round = constr |c_idx, input, output| {
+        // Add constant (weird, I thought the entire state was used here,
+        // but this is how Plonky3 does it).
+        let step_a = input[0] + INTERNAL_ROUND_CONSTANTS[c_idx];
+
+        // Apply S-box
+        let x3;
+        x3 = step_a * step_a * step_a;
+        let x7;
+        x7 = x3 * x3 * step_a;
+
+        // Multiply with the diffusion matrix
+        let line_sum = x7 + array::sum(array::sub_array(input, 1, STATE_SIZE - 1));
+        output[0] = (line_sum + DIFF_DIAGONAL[0] * x7) * DIFF_MULTIPLIER;
+        array::zip(
+            array::zip(
+                array::sub_array(input, 1, STATE_SIZE - 1),
+                array::sub_array(output, 1, STATE_SIZE - 1),
+                constr |in_v, out_v| (in_v, out_v)
+            ),
+            array::sub_array(DIFF_DIAGONAL, 1, STATE_SIZE - 1),
+            constr |(in_v, out_v), diag| out_v = (line_sum + diag * in_v) * DIFF_MULTIPLIER
+        );
+    };
+
+    // Creates a sequence of 4-byte sparsed addresses.
+    let address_inc = constr |addr_high, addr_low| {
+        let addr = array::zip(
+            addr_high,
+            addr_low,
+            |high, low| (high, low)
+        );
+
+        array::fold(
+            array::zip(
+                array::sub_array(addr, 0, array::len(addr) - 1),
+                array::sub_array(addr, 1, array::len(addr) - 1),
+                constr |(high, low), (next_high, next_low)| {
+                    increment_ptr(4, high, low, next_high, next_low)
+                }
+            ), [],
+            |a, b| a + b
+        )
+    };
+
+    // Calculate the addresses and load all the inputs into the first time step
+    let input_addr_high: col[STATE_SIZE];
+    let input_addr_low: col[STATE_SIZE];
+    address_inc(input_addr_high, input_addr_low);
+
+    let input_low: col[STATE_SIZE];
+    let input_high: col[STATE_SIZE];
+    // TODO: when link is availailable inside functions, we can turn this into array operations.
+    link if is_used ~> (input_high[0], input_low[0]) = mem.mload(input_addr_high[0], input_addr_low[0], time_step);
```

Comment:
@georgwiese all these links :grimacing:

Reply:
Don't be afraid of links ;)

In the optimized case, this should create 2 columns + some constraints for each link. But the whole machine has block size 1, so it will still be few cells to commit to. I think the main disadvantage of this approach vs larger block size & fewer links is increased verifier time (because it scales with the number of columns & constraints). We'll have to see what the tradeoff space is here.